### PR TITLE
NMS-12445: Add interfaces and common implementations for the new Graph Service API

### PR DESCRIPTION
**Pull request by mvrueden**
Here we add simple interfaces and provide some implementations for them in order to integrate with the new OpenNMS Graph Service API.

By default any graph or graph container is exposed to the Topology UI as well.
However that behaviour can be modified using the `TopologyConfiguration` interface.

Should add some documentation somewhere (-:

JIRA: https://issues.opennms.org/browse/NMS-12445

### CODE REVIEW

**Comment by mvrueden** [comment]
Build is still failing, need to take a closer look.

File: common/pom.xml [hunk 1]
```diff
@@ -14,6 +14,12 @@
             <groupId>org.opennms.integration.api</groupId>
             <artifactId>api</artifactId>
         </dependency>
+        <!-- TODO MVR can we just add this without issues? -->
```
Comment (mattixtech) [inline common/pom.xml]: I wouldn't bring in Guava. We already set a precedent for how to write the immutable implementations without any thirdparty libraries. We just make a copy of the collection and wrap it as unmodifiable.

**Comment by mattixtech** [comment]
I'll leave it up to you if you see value in this.

For the existing immutable implementations we tried to keep a consistent api where they all expose a `newBuilder`, `newBuilderFrom` and `immutableCopy`. Its a bit tedious to implement all of it but it would be nice if all of our oia provided impls were consistent. So you could look into doing that for the new types you are introducing.

We also made sure the immutable classes were deeply immutable such that all their fields were immutable and any objects contained by any of those fields (like collections/maps) were also immutable. For example see https://github.com/OpenNMS/opennms-integration-api/blob/master/common/src/main/java/org/opennms/integration/api/v1/model/immutables/ImmutableNode.java#L60. When given a builder that accepted an interface type if you take the objects directly from the builder you have no guarantee they were an immutable implementation.

**Comment by mvrueden** [comment]
I can see some benefit of doing that.
However I am not going to be working on it at the moment. 
Mainly because there is no other implementation and there won't be any for some time. There is even documentation which states that you should not implement your own thing for any model annotated interface. Besides that most of the graph objects are key value pairs. Achieving deep immutability for those is going to be hard and at the moment not worth fiddling with.

File: common/pom.xml (hunk 1 above)
Comment (mvrueden) [reply common/pom.xml]: Yeah, good point. Originally I added it to get the original code working. But guava isn't really needed anymore.

**Comment by mvrueden** [comment]
I was probably not clear enough:
It makes sense to have a similar structure and behaviour for all model objects. So I am going to add `newBuilder`, `newBuilderFrom` and `immutableCopy` to all immutables. Will also add according methods to `ImmutableVertexRef`, `ImmutableGraphContainerInfo` and `ImmutableGraphInfo`. However I am not going to fiddle with deep immutability for the key value pairs on all graph elements.

File: api/src/main/java/org/opennms/integration/api/v1/graph/Edge.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+import org.opennms.integration.api.v1.annotations.Model;
+
+@Model
+public interface Edge {
+    String getNamespace();
+    String getId();
+    VertexRef getSource();
+    VertexRef getTarget();
+    Map<String, Object> getProperties();
```
Comment (mvrueden) [inline api/src/main/java/org/opennms/integration/api/v1/graph/Edge.java]: Would be nice to have a method which just allows to fetch a certain property without fetching all.

Basically `edge.getProperties().get("xyz")` vs. `edge.getProperty("xyz")`.

Should be done for all elements (graph, vertex, edge, container).

File: sample/src/main/java/org/opennms/integration/api/sample/MyCustomStatusProvider.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+public class MyCustomStatusProvider implements LegacyStatusProvider {
+
+    @Override
+    public boolean canCalculate(String namespace) {
+        return Lists.newArrayList(NAMESPACE_1, NAMESPACE_2).contains(namespace);
+    }
+
+    @Override
+    public StatusInfo calculateStatus(Vertex vertex) {
+        if (vertex.getNamespace().equals(NAMESPACE_1)) {
```
Comment (mvrueden) [inline sample/src/main/java/org/opennms/integration/api/sample/MyCustomStatusProvider.java]: We should add an example to actually fetch the maximum alarm associated with a certain node to have a meaningful and helpful example.

**Comment by mvrueden** [comment]
I just realize, that all exposed Graphs or GrahContainers are cached forever. I am adding something to better control the caching strategy.

File: api/src/main/java/org/opennms/integration/api/v1/dao/AlarmDao.java [hunk 4]
```diff
@@ -45,4 +47,5 @@
 
     List<Alarm> getAlarms();
 
+    Optional<Alarm> getWorstAlarm(NodeRef nodeRef);
```
Comment (j-white) [inline api/src/main/java/org/opennms/integration/api/v1/dao/AlarmDao.java]:
```suggestion
    Optional<Alarm> getAlarmWithHighestSeverity(NodeRef nodeRef);
```

Comment (mvrueden) [reply api/src/main/java/org/opennms/integration/api/v1/dao/AlarmDao.java]: Makes sense 👍

**Pull request merged**